### DISCUSSION TITLE
(Virtualbox, BIOS, maybe a beginner question) How to debug hang/black screen after telling ZBM to boot?

### DISCUSSION THREAD

**Discussion by autumn-birds** [zbm-dev/zfsbootmenu#342] [Questions] [answered]
Hi,

I'm trying to get ZFSBootMenu working in a VirtualBox VM on a Windows host.  This is because I want to transition into using Linux as a main system again, but I want a way to roll back easily (hence looking into this tool), and am also trying to get a working system in a "test environment" first before I start changing things on bare metal.

The VM is in BIOS mode, which I think is the Virtualbox default.  I have an Arch system "installed" on a dataset, with locally compiled AUR `zfs-linux-lts` and `linux-lts` as the ZFS/kernel.  I am starting ZFSBootMenu from a separate partition with syslinux, with this config:

```
DEFAULT zbm-2.0.0

LABEL zbm-2.0.0
KERNEL /vmlinuz-bootmenu
INITRD /initramfs-bootmenu.img
```

(These files are just from the *Releases* area. v2.0.0 specifically.)

ZFSBootMenu starts fine and can see my dataset.  However when I press <return> to boot into that dataset I get a blank screen with a cursor that blinks a couple of times then freezes, and the entire system seems to hang.

I've tried adding `debug nomodeset` to the KCL in zbootmenu and then booting but it didn't change anything.  I did get it to give a completely black screen one time but that wasn't much of an improvement.  I also tried enabling `zfs.target` inside the dataset chroot using the Arch installer.  The initramfs is theoretically generated by `mkinitcpio -P` in the chroot, after altering /etc/mkinitcpio.conf to have `HOOKS=(base udev autodetect modconf block keyboard zfs filesystems fsck)` rather than the default.

I'm really not sure how to try to figure out what's going on here, though I might try making a Void Linux dataset to see if that works.  Sorry if this is the wrong venue for this question, but what options do I have for trying to debug this?  Or am I just attempting to use a configuration that's not supported and simply doesn't work (and if so, what's wrong with that configuration?)

**Comment by ahesford** [comment, marked as answer]
I would first try booting your Arch kernel and initramfs directly with syslinux. You can set this up from within ZFSBootMenu: in the main menu, with your Arch environment selected, press CTRL+J to chroot into the environment. (It will be read-only, but that won't matter. If you ever need read-write access, just press CTRL+W in the menu to reimport with write privilege the pool holding the currently selected boot environment.) Mount the syslinux partition, copy your Arch kernel and initramfs into it, and add an entry to the syslinux config. Remember to add the right `root=` argument to the new entry.

If your Arch kernel boots fine directly, I would suspect that kexec doesn't play nicely with VBox. Trying a manual kexec from the ZBM emergency shell should be the next step.

**Reply by autumn-birds** [reply]
I think you're right that kexec() might not be playing nice.  Or, VirtualBox might not be playing nice with kexec().  Who knows.  It loaded fine in the VM when directly booted like you suggest, though, for some reason I didn't think of testing like that -- thanks.

I also have since tried a relatively similar setup (although in UEFI mode) on real hardware and it seems to survive the ZBM->kexec()->Arch handoff fine and boot a working system. (Working...ish, but it's alot closer than I was!)  Thanks for the help :)
